### PR TITLE
Interface update

### PR DESCRIPTION
**Pull request by FelixFan1992**

### CODE REVIEW

File: core/services/ocr2/plugins/ocr2keeper/evm21/log_event_provider.go [hunk 1]
```diff
@@ -241,7 +243,7 @@ func (p *logEventProvider) GetLogs() ([]ocr2keepers.UpkeepPayload, error) {
 		log := l.log
 		logExtension := fmt.Sprintf("%s:%d", log.TxHash.Hex(), uint(log.LogIndex))
 		trig := ocr2keepers.NewTrigger(log.BlockNumber, log.BlockHash.Hex(), logExtension)
-		payload := ocr2keepers.NewUpkeepPayload(l.id, int(logTrigger), trig, log.Data)
+		payload := ocr2keepers.NewUpkeepPayload(l.id, int(logTrigger), ocr2keepers.BlockKey(strconv.FormatInt(log.BlockNumber, 10)), trig, log.Data)
```
Comment (FelixFan1992) [inline core/services/ocr2/plugins/ocr2keeper/evm21/log_event_provider.go:246]: is this correct? @EasterTheBunny

File: core/services/ocr2/plugins/ocr2keeper/evm21/feed_lookup.go [hunk 2]
```diff
@@ -68,35 +69,37 @@ type AdminOffchainConfig struct {
...
+		ext := res.Extension.(EVMAutomationResultExtension21)
+		if ext.FailureReason != UPKEEP_FAILURE_REASON_TARGET_CHECK_REVERTED {
 			continue
 		}
 
-		block := upkeepResults[i].Block
-		upkeepId := upkeepResults[i].ID
-		opts, err := r.buildCallOpts(ctx, big.NewInt(int64(block)))
+		// in feed lookup, use the block number at the block where the log was emitted
+		block := res.Payload.Trigger.BlockNumber
```
Comment (FelixFan1992) [inline core/services/ocr2/plugins/ocr2keeper/evm21/feed_lookup.go:81]: use the block number in the trigger for feed lookup step

File: core/services/ocr2/plugins/ocr2keeper/evm21/encoder.go [hunk 3]
```diff
@@ -107,7 +108,7 @@ func (enc EVMAutomationEncoder21) Encode(results ...ocr2keepers.CheckResult) ([]
 		}
 
 		ids[i] = id
-		gasLimits[i] = new(big.Int).SetUint64(result.GasUsed)
+		gasLimits[i] = new(big.Int).SetUint64(result.GasAllocated)
```
Comment (FelixFan1992) [inline core/services/ocr2/plugins/ocr2keeper/evm21/encoder.go:111]: gas allocated as gas limit

File: core/services/ocr2/plugins/ocr2keeper/evm21/registry.go [hunk 4]
```diff
@@ -694,7 +684,8 @@ func (r *EvmRegistry) checkUpkeeps(ctx context.Context, keys []ocr2keepers.Upkee
 		var payload []byte
 		switch getUpkeepType(upkeepId.Bytes()) {
 		case logTrigger:
-			payload, err = r.abi.Pack("checkUpkeep", upkeepId, []byte{}) // TODO: pass log data
+			// check data will include the log trigger config
+			payload, err = r.abi.Pack("checkUpkeep", upkeepId, key.CheckData)
```
Comment (FelixFan1992) [inline core/services/ocr2/plugins/ocr2keeper/evm21/registry.go:688]: pass the check data to check upkeep

File: core/services/ocr2/plugins/ocr2keeper/evm21/registry.go [hunk 5]
```diff
@@ -748,25 +739,30 @@ func (r *EvmRegistry) checkUpkeeps(ctx context.Context, keys []ocr2keepers.Upkee
...
 			continue
 		}
 
-		opts, err := r.buildCallOpts(ctx, big.NewInt(int64(checkResult.Block)))
+		block, upkeepId, err := r.getCheckBlockAndUpkeepId(cr.Payload)
+		if err != nil {
+			return nil, err
+		}
+
+		opts, err := r.buildCallOpts(ctx, block)
```
Comment (FelixFan1992) [inline core/services/ocr2/plugins/ocr2keeper/evm21/registry.go:759]: simulate at the check block (NOT the trigger block)

File: core/services/ocr2/plugins/ocr2keeper/evm21/log_event_provider.go (hunk 1 above)
Comment (amirylm) [reply core/services/ocr2/plugins/ocr2keeper/evm21/log_event_provider.go:246]: I'll fix that on the log trigger branch

File: core/services/ocr2/plugins/ocr2keeper/evm21/registry.go [hunk 6]
```diff
@@ -218,16 +218,16 @@ func (r *EvmRegistry) GetActiveUpkeepIDsByType(ctx context.Context, triggers ...
 	return keys, nil
 }
 
-// TODO: should be called with ocr2keepers.UpkeepPayload
-func (r *EvmRegistry) CheckUpkeep(ctx context.Context, mercuryEnabled bool, keys ...ocr2keepers.UpkeepKey) ([]ocr2keepers.UpkeepResult, error) {
+// in upkeep payload, the trigger block number and hash are the number and hash for the block this log was emitted
+func (r *EvmRegistry) CheckUpkeep(ctx context.Context /* mercuryEnabled bool, */, keys ...ocr2keepers.UpkeepPayload) ([]ocr2keepers.CheckResult, error) {
```
Comment (FelixFan1992) [inline core/services/ocr2/plugins/ocr2keeper/evm21/registry.go:222]: who creates keys and who populates the check data?

File: core/services/ocr2/plugins/ocr2keeper/evm21/encoder.go [hunk 7]
```diff
@@ -123,6 +124,86 @@ func (enc EVMAutomationEncoder21) Encode(results ...ocr2keepers.CheckResult) ([]
...
+	// if gasLimits, ok = m[mKeys[3]].([]*big.Int); !ok {
+	// 	return res, fmt.Errorf("gas limits as wrong type")
+	// }
+
+	if performs, ok = m[mKeys[5]].([][]byte); !ok {
+		return res, fmt.Errorf("perform datas as wrong type")
+	}
+
+	for i, upkeepId := range upkeepIds {
+		payload := ocr2keepers.NewUpkeepPayload(
```
Comment (FelixFan1992) [inline core/services/ocr2/plugins/ocr2keeper/evm21/encoder.go:188]: currently i am:
1. leaving the block and check data empty
2. assuming all upkeeps are log triggered
3. cast [32]byte triggers[i].BlockHash to a string for block hash

File: core/services/ocr2/plugins/ocr2keeper/evm21/encoder.go [hunk 8]
```diff
@@ -123,6 +124,86 @@ func (enc EVMAutomationEncoder21) Encode(results ...ocr2keepers.CheckResult) ([]
...
+		return res, fmt.Errorf("upkeep ids of incorrect type in report")
+	}
+
+	// TODO: a type assertion on `wrappedTrigger` did not work, even with the
+	// exact same struct definition as what follows. reflect was used to get the
+	// struct definition. not sure yet how to clean this up.
+	// ex:
+	// t := reflect.TypeOf(rawPerforms)
+	// fmt.Printf("%v\n", t)
+	triggers, ok := m[mKeys[4]].([]struct {
```
Comment (FelixFan1992) [inline core/services/ocr2/plugins/ocr2keeper/evm21/encoder.go:160]: @amirylm you will add 2 more fields here?

Comment (EasterTheBunny) [reply core/services/ocr2/plugins/ocr2keeper/evm21/encoder.go:160]: You should be able to use the struct from the contract to derive the fields to use. Looks like the field names are different. I will need to change mine too.

```
  struct LogTrigger {
    bytes32 txHash;
    uint32 logIndex;
    uint32 blockNum;
    bytes32 blockHash;
  }
```

File: core/services/ocr2/plugins/ocr2keeper/evm21/encoder.go (hunk 7 above)
Comment (EasterTheBunny) [reply core/services/ocr2/plugins/ocr2keeper/evm21/encoder.go:188]: The trigger will need extra data in the `Extension` property. I think Amir created a `NewTrigger` function in the keepers repo to help construct this trigger from log data. You can look at the log event provider for how to use that function.

**Review by EasterTheBunny — CHANGES_REQUESTED** [review]
This is pretty much what I had in mind. Some details on the trigger decoding and struct building though.

File: core/services/ocr2/plugins/ocr2keeper/evm21/registry.go [hunk 9]
```diff
@@ -647,45 +646,35 @@ func (r *EvmRegistry) doCheck(ctx context.Context, mercuryEnabled bool, keys []o
...
-
-	parts := strings.Split(string(key), separator)
-	if len(parts) != 2 {
-		return nil, nil, fmt.Errorf("unsplittable key")
-	}
-
-	if block, ok = new(big.Int).SetString(parts[0], 10); !ok {
-		return nil, nil, fmt.Errorf("could not get block from key")
+func (r *EvmRegistry) getCheckBlockAndUpkeepId(key ocr2keepers.UpkeepPayload) (*big.Int, *big.Int, error) {
+	block, success := new(big.Int).SetString(string(key.CheckBlock), 10)
```
Comment (EasterTheBunny) [inline core/services/ocr2/plugins/ocr2keeper/evm21/registry.go:654]: I talked to Ryan and we can use the BlockNumber in the Trigger here. We shouldn't actually need `CheckBlock`.

File: core/services/ocr2/plugins/ocr2keeper/evm21/registry.go (hunk 5 above)
Comment (EasterTheBunny) [reply core/services/ocr2/plugins/ocr2keeper/evm21/registry.go:758]: I talked to Ryan and we concluded they are the same thing now. The block number in the `Trigger` is the block to run all pipeline check on.

**Review by EasterTheBunny — CHANGES_REQUESTED** [review]

**Review by amirylm — APPROVED** [review]

**Pull request merged**